orca/producer_test.go

```diff
@@ -0,0 +1,538 @@
+	orcaSrv.SetUtilization("mary", 321)
+	loadReportWant = &v3orcapb.OrcaLoadReport{
+		CpuUtilization: 50,
+		MemUtilization: 200,
+		Utilization:    map[string]float64{"bob": 555, "mary": 321},
+	}
+
+	for {
+		select {
+		case r := <-oobLis.loadReportCh:
+			t.Log("Load report received: ", r)
+			if proto.Equal(r, loadReportWant) {
+				// Success!
+				return
+			}
+		case <-ctx.Done():
+			t.Fatalf("timed out waiting for load report of: %v", loadReportWant)
+		}
+	}
+}
+
+// fakeORCAService is a simple implementation of an orca service that allows us
+// to verify requests and send responses on demand.
+type fakeORCAService struct {
+	v3orcaservicegrpc.UnimplementedOpenRcaServiceServer
+
+	reqCh  chan *v3orcaservicepb.OrcaLoadReportRequest
+	respCh chan interface{} // either *v3orcapb.OrcaLoadReport or error
+}
+
+func newFakeORCAService() *fakeORCAService {
+	return &fakeORCAService{
+		reqCh:  make(chan *v3orcaservicepb.OrcaLoadReportRequest),
+		respCh: make(chan interface{}),
+	}
+}
+
+func (f *fakeORCAService) close() {
+	close(f.respCh)
+}
+
+func (f *fakeORCAService) StreamCoreMetrics(req *v3orcaservicepb.OrcaLoadReportRequest, stream v3orcaservicegrpc.OpenRcaService_StreamCoreMetricsServer) error {
+	f.reqCh <- req
+	for resp := range f.respCh {
+		if err, ok := resp.(error); ok {
+			return err
+		}
+		if err := stream.Send(resp.(*v3orcapb.OrcaLoadReport)); err != nil {
+			// In the event that a stream error occurs, a new stream will have
+			// been created that was waiting for this response message.  Push
+			// it back onto the channel and return.
+			//
+			// This happens because we range over respCh.  If we changed to
+			// instead select on respCh + stream.Context(), the same situation
+			// could still occur due to a race between noticing the two events,
+			// so such a workaround would still be needed to prevent flakiness.
+			f.respCh <- resp
+			return err
+		}
+	}
+	return nil
+}
+
+// TestProducerBackoff verifies that the ORCA producer applies the proper
+// backoff after stream failures.
+func (s) TestProducerBackoff(t *testing.T) {
+	grpctest.TLogger.ExpectErrorN("injected error", 4)
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	// Provide a convenient way to expect backoff calls and return a minimal
+	// value.
+	expectedBackoff := -1 // -1 indicates any value is allowed.
+	oldBackoff := internal.DefaultBackoffFunc
+	internal.DefaultBackoffFunc = func(got int) time.Duration {
+		if expectedBackoff != -1 {
+			if got != expectedBackoff {
+				t.Errorf("Unexpected backoff received; got %v want %v", got, expectedBackoff)
+			}
+		}
+		return time.Millisecond
+	}
+	defer func() { internal.DefaultBackoffFunc = oldBackoff }()
+
+	// Initialize listener for our ORCA server.
+	lis, err := testutils.LocalTCPListener()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Register our fake ORCA service
+	s := grpc.NewServer()
+	fake := newFakeORCAService()
```

Comment:
Can you rename this local var? fakeWhat? I was having trouble later in the test figuring out what this was a fake of.

Reply:
fos?

Reply:
That's far less intuitive to me.